provider/provider.go

```diff
@@ -0,0 +1,270 @@
+	// TypeName is the name of the resource type to plan.
+	TypeName string
+
+	// PriorState is the previously saved state value for this resource.
+	PriorState cty.Value
+
+	// ProposedNewState is the expected state after the new configuration is
+	// applied. This is created by directly applying the configuration to the
+	// PriorState. The provider is then responsible for applying any further
+	// changes required to create the proposed final state.
+	ProposedNewState cty.Value
+
+	// PriorPrivate is the previously saved private data returned from the
+	// provider during the last apply.
+	PriorPrivate []byte
+}
+
+type PlanResourceChangeResponse struct {
+	// PlannedState is the expected state of the resource once the current
+	// configuration is applied.
+	PlannedState cty.Value
+
+	// RequiresReplace is the list of thee attributes that are requiring
+	// resource replacement.
+	RequiresReplace []cty.Path
+
+	// PlannedPrivate is an opaque blob that is not interpreted by terraform
+	// core. This will be saved and relayed back to the provider during
+	// ApplyResourceChange.
+	PlannedPrivate []byte
+
+	// Diagnostics contains any warnings or errors from the method call.
+	Diagnostics tfdiags.Diagnostics
+}
+
+type ApplyResourceChangeRequest struct {
+	// TypeName is the name of the resource type being applied.
+	TypeName string
+
+	// PriorState is the current state of resource.
+	PriorState cty.Value
+
+	// Planned state is the state returned from PlanResourceChange, and should
+	// represent the new state, minus any remaining computed attributes.
+	PlannedState cty.Value
+
+	// PlannedPrivate is the same value as returned by PlanResourceChange.
+	PlannedPrivate []byte
+}
+
+type ApplyResourceChangeResponse struct {
+	// NewState is the new complete state after applying the planned change.
+	// In the event of an error, NewState should represent the most recent
+	// known state of the resource, if it exists.
+	NewState cty.Value
+
+	// Connection is a map of string values used to return any information
+	// provisioners might require to connect to the resource.
+	Connection map[string]string
+
+	// Private is an opaque blob that will be stored in state along with the
+	// resource. It is intended only for interpretation by the provider itself.
+	Private []byte
+
+	// Diagnostics contains any warnings or errors from the method call.
+	Diagnostics tfdiags.Diagnostics
+}
+
+type ImportResourceStateRequest struct {
+	// TypeName is the name of the resource type to be imported.
+	TypeName string
+
+	// ID is a string with which the provider can identify the resource to be
+	// imported.
+	ID string
+}
+
+type ImportResourceStateResponse struct {
+	// State contains one or more state values for the imported resource. It is
+	// not required that these be complete, only that there is enough
+	// identifying information for the provider to successfully update the
+	// state in ReadResource.
+	State []cty.Value
```

Comment:
The old interface was hiding something here: the provider actually needs to return a resource type alongside each of these values, which in the old interface was done by setting a special field in `InstanceState.Ephemeral` that was only used for this situation.

I had written up [`ImportedObject`](https://github.com/hashicorp/terraform/blob/4b97715bcddf67d287e49f37ac845e9558306897/states/import.go#L10-L22) to represent this situation but now seeing this in context I'd lean towards moving that struct into this package (`provider.ImportedObject`) since its only purpose is to carry the typed value from the provider into the import handler in core, where it'll then get converted into a number of "real" object states. The `states` package itself doesn't actually have any need for that type.

Reply:
Ah yes, checking back to the old interface, this is allowed to return multiple _different_ types. Do we actually have a use case for `Private` during import, since we now know the schema version ahead of time?

I agree `ImportedObject` might be better suited here.

Reply:
In general I'd assume that any object coming _from_ a provider must always be conforming to the current schema version, which we should already know out-of-band from the `GetSchema` response.

However, in writing that out I realized I missed something on first pass: `GetSchemaResponse` only contains the `configschema.Block` for each resource and not the corresponding schema version number that goes along with it. We'll need that to properly record the schema version for objects returned from both import, refresh, and apply. (note that the schema version now must be a first-class thing so that Core can see it, where before it was smuggled inside the private meta.)

Returning an initial value of `Private` in `ImportedObject` feels reasonable here too, now that you mention it. We do immediately refresh the object after we've imported it, and so in principle the provider could populate the private data there, but letting them populate it during import avoids the need for a special case in the `Read` implementation just for imported objects.